### PR TITLE
Node worker usage

### PR DESCRIPTION
**Pull request by Krizzu**
TODO:
- [x] tests are passing
- [x] `bundle` command is working
- [x] HMR is working
- [x] symbolicate is working
- [x] workers are not clogging system when something unexpected happens
- [x] redirect progress events from forks to parent process
- [x] document what the heck is going on there
- [x] fix crazy things happening with 2 compilations and express logging

### CODE REVIEW

**Comment by Krizzu** [comment]
Pointing to master now? 💪

File: integration_tests/__tests__/start.test.js [hunk 1]
```diff
@@ -18,17 +18,32 @@ const TEST_PROJECT_DIR = path.resolve(
...
+    const message = stripAnsi(data.toString()).trim();
+
+    if (message.length > 0) {
+      messageBuffer.push(message);
+    }
 
-  const { stdout } = runHaulSync(TEMP_DIR, ['start']);
-  expect(stripAnsi(stdout).trim()).toMatchSnapshot();
+    if (message.match(/ERROR/g)) {
+      done.fail(message);
```
Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: Can't we listen to Haul's events instead of matching stdout?

**Comment by thymikee** [comment]
@zamotany locally the `start.test.js` is failing for me, as there's no text output while listening to stdout (just some whitespace). Mind looking into this?

File: integration_tests/__tests__/start.test.js (hunk 1 above)
Comment (zamotany) [reply integration_tests/__tests__/start.test.js]: We can, but we need to change the way we run haul.

Comment (thymikee) [reply integration_tests/__tests__/start.test.js]: Would be great to do it as a followup. Feel free to disable this test for the time being.

File: src/compiler/Compiler.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2017-present, Callstack.
+ * All rights reserved.
+ *
+ * @flow
+ */
+
+import type { Platform, Logger } from '../types';
+
+const EventEmitter = require('events');
```
Comment (satya164) [inline src/compiler/Compiler.js]: We're using `import` in various other places. Can we change them to `require` or change these to `import` for consistency?

File: src/compiler/Compiler.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  terminate(error?: Error) {
+    this.logger.info('Shutting down Haul.');
+
+    error && this.logger.error(error.message);
+
+    Object.keys(this.forks).forEach(platform =>
+      this.forks[platform].terminate()
+    );
+    process.exit(error ? 1 : 0);
```
Comment (satya164) [inline src/compiler/Compiler.js]: Maybe we shouldn't do `process.exit` in the compiler instance, assuming that this will eventually be the basis of the public API.

File: src/compiler/TaskQueue.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    this.nextId = 0;
+  }
+
+  add(item: any) {
+    const id = this.nextId++;
+    this.queue.push({ id, item });
+    return id;
+  }
+
+  get(id: number) {
```
Comment (satya164) [inline src/compiler/TaskQueue.js]: Probably `get` isn't the best name as it also removes the item from the queue

File: src/compiler/createWebSocketServer.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+const xpipe = require('xpipe');
+const http = require('http');
+const { Server } = require('ws');
+const fs = require('fs');
+
+/**
+ * Create WebSocket server using HTTP server over Unix socket.
+ */
+module.exports = function createWebSocketServer() {
+  const socketAddress = xpipe.eq(`/tmp/HAUL_SOCKET_.socket`);
```
Comment (satya164) [inline src/compiler/createWebSocketServer.js]: Can we change it to a unique address? We should be able to run multiple instances of Haul with different ports. It's also necessary for expo integration.

File: src/compiler/createWebSocketServer.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  if (fs.existsSync(socketAddress)) {
+    fs.unlinkSync(socketAddress);
```
Comment (satya164) [inline src/compiler/createWebSocketServer.js]: We should also delete the file when the compiler terminates.

File: src/compiler/Compiler.js (hunk 2 above)
Comment (zamotany) [reply src/compiler/Compiler.js]: Does importing a type with `require` works for u? And Haul is using `requires` everywhere.

Comment (thymikee) [reply src/compiler/Compiler.js]: @satya164 let's do this in one PR with a codemod, no need to mess this now

Comment (satya164) [reply src/compiler/Compiler.js]: @zamotany no, will need import 🤔
@thymikee okay

File: src/server/util/runAdbReverse.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+module.exports = function runAdbReverse(port: number) {
+  const command = `adb reverse tcp:${port} tcp:${port}`;
+
+  exec(command, error => {
+    if (error) {
+      // Get just the error message
+      const message = error.message.split('error:')[1];
+
+      logger.warn(
```
Comment (zamotany) [inline src/server/util/runAdbReverse.js]: @thymikee @satya164 @Krizzu Do you have any idea where can we put a info about this in the UI?

File: integration_tests/utils.js [hunk 8]
```diff
@@ -109,6 +110,25 @@ function replaceTestPath(string: string) {
   return string.replace(/^\W+(.*)integration_tests/gm, '<<REPLACED>>');
 }
 
+function fetchBundle(url: string): Promise<any> {
+  return new Promise((resolve, reject) => {
```
Comment (satya164) [inline integration_tests/utils.js]: Looks like we have a dependency on `node-fetch`. Maybe worth using it rather than wrapping `http.get` in a promise?

File: src/babelRegister.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env node
```
Comment (satya164) [inline src/babelRegister.js]: Do we need this shebang? Do we ever execute this file directly?

File: src/compiler/Compiler.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    return Events;
+  }
+
+  forks: { [key: Platform]: Fork };
+  tasks: TaskQueue;
+
+  constructor(options: *) {
+    super();
+    this.forks = {};
+    this.tasks = new TaskQueue();
```
Comment (satya164) [inline src/compiler/Compiler.js]: nit: could use the class properties for initialization too

File: src/compiler/Fork.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+const createForkProcess = require('./createForkProcess');
+const createWebSocketServer = require('./createWebSocketServer');
+
+type ForkConstructorArgs = {
+  platform: Platform,
+  options: *,
+};
+
+const forks = {};
+let transportServer;
```
Comment (satya164) [inline src/compiler/Fork.js]: Is there a reason these two are not properties in the `Compiler` instance? Seems like when you create multiple compiler instances from a single process, it's going to reuse the same transport server (with the same address) and overwrite previous forks?

Will probably also be an issue if you had created and terminated a compiler instance previously since this variable is still present, so if you create a new compiler instance, it'll try to use it, but the connection has closed.

File: src/compiler/createWebSocketServer.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Create WebSocket server using HTTP server over Unix socket.
+ */
+module.exports = function createWebSocketServer() {
+  const socketAddress = getSocketAddress();
+
+  const httpServer = http.createServer();
+  const webSocketServer = new Server({ server: httpServer });
+
+  httpServer.listen(socketAddress);
```
Comment (satya164) [inline src/compiler/createWebSocketServer.js]: Seems like this `httpServer` will never close. Will it be a memory leak if you're creating and terminating multiple compilers?

File: src/compiler/worker/runWebpackCompiler.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  [key: string]: string,
+  fs: Object,
+}) {
+  const emitter = new EventEmitter();
+
+  const { configPath, configOptions } = JSON.parse(options);
+  const config = getConfig(configPath, configOptions, platform);
+
+  let lastPercent = -1;
+  config.plugins.push(
```
Comment (satya164) [inline src/compiler/worker/runWebpackCompiler.js]: Maybe clone the config instead of mutating it? If the user is sharing same config between multiple compilers, this could add multiple plugins to the same config.

File: src/compiler/worker/runWebpackCompiler.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+  emitter.on('start', () => {
+    compiler.watch({}, () => {});
+    emitter.emit(Events.BUILD_START);
+  });
+
+  return emitter;
+};
+
+function resolveAsync(args) {
```
Comment (satya164) [inline src/compiler/worker/runWebpackCompiler.js]: What does this do?

File: integration_tests/utils.js (hunk 8 above)
Comment (zamotany) [reply integration_tests/utils.js]: I'm in the middle of fixing `start` test suite, which also replaces this with `fetch`

File: src/babelRegister.js (hunk 9 above)
Comment (zamotany) [reply src/babelRegister.js]: Nope

File: src/compiler/createWebSocketServer.js (hunk 12 above)
Comment (zamotany) [reply src/compiler/createWebSocketServer.js]: It can be, though having multiple compilers is an edge case.

File: src/compiler/worker/runWebpackCompiler.js (hunk 14 above)
Comment (zamotany) [reply src/compiler/worker/runWebpackCompiler.js]: No idea. cc: @Krizzu

File: src/compiler/worker/runWebpackCompiler.js (hunk 13 above)
Comment (zamotany) [reply src/compiler/worker/runWebpackCompiler.js]: Doesn't matter - `makeReactNativeConfig` always create a fresh copy of the config.

File: .flowconfig [hunk 15]
```diff
@@ -7,5 +7,8 @@
 .*/node_modules/fbjs/*
 <PROJECT_ROOT>/fixtures/.*
 
+[untyped]
+.*/node_modules/react-stream-renderer
```
[the remark this answers is not in the document]
Comment (zamotany) [reply .flowconfig]: #no_shame

File: src/compiler/worker/runWebpackCompiler.js (hunk 14 above)
Comment (Krizzu) [reply src/compiler/worker/runWebpackCompiler.js]: If this function is fired async, second argument is `done` callback.

**Comment by zamotany** [comment]
@thymikee done

File: src/utils/makeReactNativeConfig.js [hunk 16]
```diff
@@ -60,13 +57,12 @@ const getDefaultConfig = ({
...
 }): WebpackConfig => {
   // Getting Minor version
-  const platformProgressBar = haulProgressBar(platform);
   return {
     mode: dev ? 'development' : 'production',
     context: root,
     entry: [],
     output: {
-      path: path.join(root, 'dist'),
+      path: path.join(root), // removed 'dist' from path
```
Comment (satya164) [inline src/utils/makeReactNativeConfig.js]: > `// removed 'dist' from path`

not sure what's this!!!

File: integration_tests/__tests__/start.test.js [hunk 17]
```diff
@@ -57,14 +34,48 @@ function testPlatform(platform, done) {
...
+      done();
+    }
+  }
+  _done.fail = done.fail;
+
+  testPlatform('ios', _done);
+  testPlatform('android', _done);
+});
+
+async function testPlatform(platform, done) {
```
Comment (satya164) [inline integration_tests/__tests__/start.test.js]: what do you think about changing this to return a promise instead of accepting a callback? I think tests will become simpler that way.

```js
async function testPlatform(platform) {
  const res = await fetch(`http://localhost:8081/index.${platform}.bundle`);
  const bundle = await res.text();
  expect(bundle).toMatch('__webpack_require__');
}
```

Then tests will be like:

```js
test('starts server and bundling iOS platform', () =>
  testPlatform('ios');
);
 
test('starts server and bundling Android platform', () =>
  testPlatform('android');
);
 
test('starts server and bundling all platforms', () =>
  Promise.all([
    testPlatform('ios');
    testPlatform('android');
  ]);
);
```

Comment (thymikee) [reply integration_tests/__tests__/start.test.js]: With current on-demand bundling, does `start --platform` even make sense? I think we should remove or adjust these tests

Comment (zamotany) [reply integration_tests/__tests__/start.test.js]: Test __doesn't__ spawn `haul start` with `--platform` option. `testPlatform` is only to make a request for specific platform bundle.

Comment (thymikee) [reply integration_tests/__tests__/start.test.js]: @zamotany obviously, looks like I missed that 🤦‍♂️

File: src/utils/makeReactNativeConfig.js (hunk 16 above)
Comment (zamotany) [reply src/utils/makeReactNativeConfig.js]: Not sure either

File: integration_tests/__tests__/start.test.js [hunk 18]
```diff
@@ -20,9 +20,13 @@ const TEST_PROJECT_DIR = path.resolve(
...
 let haul;
-beforeAll(done => {
+beforeAll(async done => {
   run('yarn --mutex network', TEST_PROJECT_DIR);
-  haul = runHaul(TEST_PROJECT_DIR, ['start']);
+  try {
+    haul = await runHaul(TEST_PROJECT_DIR, ['start']);
+  } catch (error) {
+    done.fail(error);
+  }
```
Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: you can get rid of the `try/catch` and `done` entirely, error will be caught normally

File: src/server/index.js [hunk 19]
```diff
@@ -41,6 +41,7 @@ const WebSocketDebuggerProxy = require('./util/WebsocketDebuggerProxy');
  */
 function createServer(config: { configPath: string, configOptions: Object }) {
   const appHandler = express();
+  appHandler.disable('etag');
```
Comment (thymikee) [inline src/server/index.js]: Why is this needed?

Comment (zamotany) [reply src/server/index.js]: to remove server caching and 304s

Comment (thymikee) [reply src/server/index.js]: worth to left a comment :)

File: src/__tests__/__snapshots__/cliEntry.test.js.snap [hunk 20]
```diff
@@ -6,12 +6,10 @@ Starts a new webpack server
 
 Options:
 
-  --port [number]                 Port to run your webpack server, 8081 by default
-  --dev [true|false]              Whether to build in development mode, true by default
-  --minify [true|false]           Whether to minify the bundle, 'true' by default when dev=false
-  --platform <ios|android|all>    Platform to bundle for
```
Comment (satya164) [inline src/__tests__/__snapshots__/cliEntry.test.js.snap]: We still support this flag for bundling, right? Maybe this should be left here?

File: src/compiler/worker/initWorker.js [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+const WebSocket = require('ws');
+const MemoryFileSystem = require('memory-fs');
+const mime = require('mime-types');
+
+const Events = global.requireWithRootDir('../events');
+const runWebpackCompiler = global.requireWithRootDir('./runWebpackCompiler');
+
+module.exports = function initWorker({
+  platform,
+  // fileOutput,
```
Comment (satya164) [inline src/compiler/worker/initWorker.js]: Commented out code

File: src/__tests__/__snapshots__/cliEntry.test.js.snap (hunk 20 above)
Comment (zamotany) [reply src/__tests__/__snapshots__/cliEntry.test.js.snap]: This is `start` command snapshot not `bundle`

File: src/compiler/Compiler.js [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+      if (!Object.keys(this.forks).length) {
+        callback({
+          errors: null,
+          platform: null,
+          file: null,
+          mimeType: null,
+        });
+      }
+
+      // Callback will be invoked on `FILE_RECEIVED` event or on `FILE_NOT_FOUND`.
```
Comment (thymikee) [inline src/compiler/Compiler.js]: is this comment still relevant here?

File: src/compiler/Fork.js [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+    );
+
+    forks[platform] = this;
+  }
+
+  setSocket(socket: WebSocket) {
+    this.socket = socket;
+
+    // Flush enqueued messages.
+    this.enqueuedMessages.forEach(({ type, ...payload }) =>
```
Comment (thymikee) [inline src/compiler/Fork.js]: Could use `.filter` instead of `.forEach` and just return `false` each time, but meh

File: src/compiler/Compiler.js (hunk 22 above)
Comment (zamotany) [reply src/compiler/Compiler.js]: Yes

File: src/compiler/Fork.js [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+      this.send(type, payload)
+    );
+    this.enqueuedMessages = [];
+
+    // $FlowFixMe
+    this.socket.addEventListener('message', ({ data }) => {
+      const { type, ...payload } = JSON.parse(data.toString());
```
Comment (thymikee) [inline src/compiler/Fork.js]: we could guard that parsing somehow, in case data is undefined?

Comment (zamotany) [reply src/compiler/Fork.js]: We could but if data is undefined the process should terminate.

File: src/server/middleware/symbolicateMiddleware.js [hunk 25]
```diff
@@ -109,32 +103,48 @@ function getRequestedFrames(req: $Request): ?ReactNativeStack {
...
+    const bundleName = `${name}.${platform}.${rest[rest.length - 1]}`;
 
     // grab our source map consumer
-    const consumer = createSourceMapConsumer(compiler);
+    const consumer = await createSourceMapConsumer(
+      compiler,
+      // $FlowFixMe
+      `http://localhost:${req.get('host').split(':')[1]}/${bundleName}.map`
+    );
+    // console.log('C', consumer);
```
Comment (thymikee) [inline src/server/middleware/symbolicateMiddleware.js]: dead comment

File: src/server/middleware/symbolicateMiddleware.js [hunk 26]
```diff
@@ -109,32 +103,48 @@ function getRequestedFrames(req: $Request): ?ReactNativeStack {
...
+    }
+
+    const [name, ...rest] = filename.split('.');
+    const bundleName = `${name}.${platform}.${rest[rest.length - 1]}`;
 
     // grab our source map consumer
-    const consumer = createSourceMapConsumer(compiler);
+    const consumer = await createSourceMapConsumer(
+      compiler,
+      // $FlowFixMe
```
Comment (thymikee) [inline src/server/middleware/symbolicateMiddleware.js]: is this needed? line below is a string

**Comment by Krizzu** [comment]
@zamotany 👏 !
